### PR TITLE
Add status code and response to server errors

### PR DESCRIPTION
**Pull request by ashkan18**
# Problem
Currently when `Graphlient` errors are raised, we do see useful `message` but there are cases that we want to get more details about the error. GraphQL specs allow for other fields in an error response like `extensions` which can have application specific fields:
http://facebook.github.io/graphql/June2018/#example-fce18

Current setup doesn't allow easy access to error details.

# Solution
Add `response` to `Graphlient::Errors::Error` to make sure all the errors raised in Graphlient's level have the response returned from the server in them. 
Update `Graphlient::Errors::ServerError` to also include/expose `status_code` along with the response.

This way clients can access the details of the raised error if they wanted to.

### CODE REVIEW

**Comment by dblock** [comment]
This looks like an interface change. Can we keep `ServerError`'s constructor the same signature and just add the response part?

**Comment by yuki24** [comment]
Yeah we should avoid breaking changes whenever possible. Once exposed, people will depend on a surprising mount of things and upgrading would become real hard if a path to a new interface is not given. I also personally don't mind chaining calls like `server_error.cause.response[:status]`.

**Comment by ashkan18** [comment]
good points, will update 🙌

**Comment by ashkan18** [comment]
now i remember why i did it this way, depending on the http Adapter being used we may have different object in `inner_exception` which makes it little hard to parse, in case of `Faraday`, `inner_exception` is an actual `Faraday::ClientError` but in case of `HttpAdapter` is a `response`. 
I think there is benefit in easy access to response and status, I see two options:
1) Have `ServerError` get `inner_exception, status_code, response_body` on initialize
```ruby
module Graphlient
  module Errors
    class ServerError < Error
      attr_reader :status_code, :inner_exception
      def initialize(message, inner_exception, http_status, response_body)
        super(message, response_body)
        @status_code = http_status
        @inner_exception = inner_exception
      end
    end
  end
end
```

2) Have conditional code in `ServerError` that based on what it got from `inner_exception` extracts the response body and status, something like
```ruby
def status_code
  if inner_exception.is_a?(Faraday::ClientError)
     inner_exception.response.status
  elsif inner_exception.responds_to? :status
     inner_exception.status
  end
end
```

I personally prefer option #1 but let me know if you think otherwise or there are other options.

**Comment by dblock** [comment]
What about inheriting `Faraday::ServerError` from `ServerError` and doing faraday-specific stuff in that?

**Comment by ashkan18** [comment]
ooh, thats interesting idea! giving it a try!

File: README.md [hunk 1]
```diff
@@ -123,7 +123,10 @@ Unlike graphql-client, Graphlient will always raise an exception unless the quer
 * [Graphlient::Errors::ClientError](lib/graphlient/errors/client_error.rb): all client-side query validation failures based on current schema
 * [Graphlient::Errors::GraphQLError](lib/graphlient/errors/graphql_error.rb): all GraphQL API errors, with a humanly readable collection of problems
 * [Graphlient::Errors::ExecutionError](lib/graphlient/errors/execution_error.rb): all GraphQL execution errors, with a humanly readable collection of problems
-* [Graphlient::Errors::ServerError](lib/graphlient/errors/server_error.rb): all transport errors raised by Faraday
+* [Graphlient::Errors::ServerError](lib/graphlient/errors/server_error.rb): all transport errors raised by Http Adapters. You can access `inner_exception`, `status_code` and `response` on these errors to get more details on what went wrong
```
Comment (dblock) [inline README.md]: Http -> HTTP

File: README.md [hunk 2]
```diff
@@ -123,7 +123,10 @@ Unlike graphql-client, Graphlient will always raise an exception unless the quer
...
+* [Graphlient::Errors::FaradayServerError](lib/graphlient/errors/faraday_server_error.rb): This inherits from `ServerError` ☝️, we recommend using `ServerError` to catch these.
```
Comment (dblock) [inline README.md]: Either `This ... .` everywhere or `this` without a period. Make it consistent with above.

Comment (dblock) [reply README.md]: In Ruby we say "rescue" I think instead of "catch"

File: lib/graphlient/errors/error.rb [hunk 3]
```diff
@@ -1,13 +1,6 @@
 module Graphlient
   module Errors
     class Error < StandardError
-      attr_reader :inner_exception
```
Comment (dblock) [inline lib/graphlient/errors/error.rb]: Is there any `Error` raised anywhere without an `inner_exception`? If not, I would undo this and make sure to pass it through, but otherwise this change is fine.

Comment (yuki24) [reply lib/graphlient/errors/error.rb]: You could always use [the `exception#cause` method](http://www.virtuouscode.com/2013/12/25/exception-causes-in-ruby-2-1/) to get the cause. While it's implicit, It is the way if the user of this gem wants to know if there's any exception raised eariler. I would leave it as it is and send a separate PR to deprecate it for compatibility.

Comment (yuki24) [reply lib/graphlient/errors/error.rb]: You could always use [the `exception#cause` method](http://www.virtuouscode.com/2013/12/25/exception-causes-in-ruby-2-1/) to get the cause. While it's implicit, It is the way if the user of this gem wants to know if there's any exception raised eariler. I would leave it as it is and send a separate PR to deprecate it for compatibility.

Comment (dblock) [reply lib/graphlient/errors/error.rb]: Nice! I haven't been using that, it seems to exist since Ruby 2.1.

Comment (ashkan18) [reply lib/graphlient/errors/error.rb]: Nice! didn't know about that, I agree with switching to it in a separate PR 🙌

File: README.md [hunk 4]
```diff
@@ -123,7 +123,10 @@ Unlike graphql-client, Graphlient will always raise an exception unless the quer
 * [Graphlient::Errors::ClientError](lib/graphlient/errors/client_error.rb): all client-side query validation failures based on current schema
 * [Graphlient::Errors::GraphQLError](lib/graphlient/errors/graphql_error.rb): all GraphQL API errors, with a humanly readable collection of problems
 * [Graphlient::Errors::ExecutionError](lib/graphlient/errors/execution_error.rb): all GraphQL execution errors, with a humanly readable collection of problems
-* [Graphlient::Errors::ServerError](lib/graphlient/errors/server_error.rb): all transport errors raised by Faraday
+* [Graphlient::Errors::ServerError](lib/graphlient/errors/server_error.rb): all transport errors raised by HTTP Adapters. You can access `inner_exception`, `status_code` and `response` on these errors to get more details on what went wrong
+* [Graphlient::Errors::FaradayServerError](lib/graphlient/errors/faraday_server_error.rb): this inherits from `ServerError` ☝️, we recommend using `ServerError` to rescue these.
```
Comment (dblock) [inline README.md]: I know I'm OCD, but this still has a period at the end.

File: lib/graphlient/errors/execution_error.rb [hunk 5]
```diff
@@ -4,12 +4,12 @@ class ExecutionError < Error
...
 
       def initialize(response)
+        super('the server responded with a GraphQL execution error')
         @response = response
-        super 'the server responded with a GraphQL execution error'
       end
 
       def errors
-        response.data.errors
+        @response.data.errors
```
Comment (dblock) [inline lib/graphlient/errors/execution_error.rb]: This shouldn't be `@`, we allow overriding things in general, and this would prevent that.

File: README.md (hunk 4 above)
Comment (ashkan18) [reply README.md]: hahaha a PR is probably the only place OCD is recommended... 🙏

**Comment by ashkan18** [comment]
Thanks both for reviews, going to squash merge 🙌
